### PR TITLE
Add Test requirement for using local Refresh from Network server.

### PR DESCRIPTION
**Pull request by mikeytdisco**
# Preface

## Description

This PR adds a requirement to use a local Refresh from Network server, and how to set one up on a Windows system using WSL2 and Docket Desktop.  This PR is a test environment only PR, and does not affect Dfci operation.

For each item, place an "x" in between `[` and `]` if true. Example: `[x]`.
_(you can also check items in the GitHub UI)_

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [x] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [x] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

Tested locally

## Integration Instructions

The ability to test Refresh from Network has changed significantly.
You will have to read the [DfciTests Documentation](UnitTests/DfciTests/readme.md#setting-up-the-host-system-with-the-refresh-from-network-server-running-as-a-container), as there are significant changes on testing Refresh from Network.
Instead of a hand built Azure server, each entity testing Dfci has to publish their own Refresh Server.

### CODE REVIEW

File: DfciPkg/UnitTests/DfciTests/Support/Python/DFCI_SupportLib.py [hunk 1]
```diff
@@ -9,17 +9,24 @@
...
-    from StringIO import StringIO
-except ImportError:
-    from io import StringIO, BytesIO
+
+# try:
+# from StringIO import StringIO
+# except ImportError:
+from io import StringIO, BytesIO
+# from cryptography.hazmat.primitives.asymmetric import rsa
+# from cryptography.hazmat.primitives.serialization import load_der_public_key
```
Comment (os-d) [inline DfciPkg/UnitTests/DfciTests/Support/Python/DFCI_SupportLib.py:29]: Do we want to keep the commented out code or can we remove it?

**Review by os-d — APPROVED** [review]

File: DfciPkg/Application/DfciMenu/DfciCheck429.c [hunk 2]
```diff
@@ -0,0 +1,72 @@
+/** @file
+DfciSARecovery.c
+
+Device Firmware Configuration Interface - Stand Alone driver that can be loaded
```
Comment (apop5) [inline DfciPkg/Application/DfciMenu/DfciCheck429.c:4]: File header needs updated.

File: DfciPkg/Application/DfciMenu/DfciCheck429.c [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+#include <Library/UefiLib.h>
+
+#include "DfciPrivate.h"
+#include "DfciRequest.h"
+
+// *---------------------------------------------------------------------------------------*
+// * Application Global Variables                                                          *
+// *---------------------------------------------------------------------------------------*
+DFCI_NETWORK_REQUEST  mDfciNetworkRequest = { 0 };
+STATIC CHAR8          mCheck429Url[]      = "http://mikeytbds3.eastus.cloudapp.azure.com/return_429";
```
Comment (apop5) [inline DfciPkg/Application/DfciMenu/DfciCheck429.c:30]: Do you want this URL checked in?

File: DfciPkg/Application/DfciMenu/DfciCheck429.c [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+DfciCheck429Entry (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS  Status;
+
+  ZeroMem (&mDfciNetworkRequest, sizeof (mDfciNetworkRequest));
+
+  mDfciNetworkRequest.HttpRequest.Url = AllocateCopyPool (sizeof (mCheck429Url), &mCheck429Url);
```
Comment (apop5) [inline DfciPkg/Application/DfciMenu/DfciCheck429.c:50]: Can you verify that the AllocateCopyPool gets memory?

File: DfciPkg/Application/DfciMenu/DfciCheck429.c [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  mDfciNetworkRequest.MainLogic       = Check429Logic;
+
+  // Try every NIC in the system until one fills the first part of the request.
+  Status = TryEachNICThenProcessRequest (&mDfciNetworkRequest);
+
+  DEBUG ((DEBUG_INFO, "Url   %a\n", mDfciNetworkRequest.HttpRequest.Url));
+  DEBUG ((DEBUG_INFO, "HttpStatus = %d, Status-%r\n", mDfciNetworkRequest.HttpStatus.HttpStatus, Status));
+  DEBUG ((DEBUG_INFO, "HttpStatus = %a\n", GetHttpErrorMsg (mDfciNetworkRequest.HttpStatus.HttpStatus)));
+
```
Comment (apop5) [inline DfciPkg/Application/DfciMenu/DfciCheck429.c:59]: Should we check that TryEachNICThenProcessRequest returned success before we access the data?

**Review by apop5 — APPROVED** [review]

File: DfciPkg/Application/DfciMenu/DfciCheck429.inf [hunk 6]
```diff
@@ -0,0 +1,69 @@
+## @file
+# DfciSARecovery.inf
```
Comment (spbrogan) [inline DfciPkg/Application/DfciMenu/DfciCheck429.inf:2]: incorrect

**Review by spbrogan — APPROVED** [review]

File: DfciPkg/Application/DfciMenu/DfciRequest.c [hunk 7]
```diff
@@ -818,16 +817,16 @@ DfciBuildRequestHeaders (
   RequestHeaders[1].FieldValue = AllocateCopyPool (AsciiStrSize (HEADER_AGENT_VALUE), HEADER_AGENT_VALUE);
   RequestHeaders[2].FieldName  = AllocateCopyPool (AsciiStrSize (HTTP_HEADER_ACCEPT), HTTP_HEADER_ACCEPT);
   RequestHeaders[2].FieldValue = AllocateCopyPool (AsciiStrSize (HEADER_ACCEPT_VALUE), HEADER_ACCEPT_VALUE);
+  RequestHeaders[3].FieldName  = AllocateCopyPool (AsciiStrSize (HTTP_HEADER_CONTENT_LENGTH), HTTP_HEADER_CONTENT_LENGTH);
+  AsciiSPrint (ContentLengthString, sizeof (ContentLengthString), "%ld", BodyLength);
+  RequestHeaders[3].FieldValue = AllocateCopyPool (AsciiStrSize (ContentLengthString), ContentLengthString);
```
[the remark this answers is not in the document]
Comment (mikeytdisco) [reply DfciPkg/Application/DfciMenu/DfciRequest.c:822]: Fixed

File: DfciPkg/Application/DfciMenu/DfciCheck429.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (mikeytdisco) [reply DfciPkg/Application/DfciMenu/DfciCheck429.c]: Fixed.

File: DfciPkg/UnitTests/DfciTests/CloneUnitTests.bat [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+git init .
+git remote add mu_feature_dfci https://github.com/Microsoft/mu_feature_dfci
+git fetch mu_feature_dfci
+git checkout mu_feature_dfci/main -- DfciPkg/UnitTests/DfciTests
+rmdir /s /q .git
+goto Done
+
+:Error
+echo Don't run this command in a direcotry at the root of a repository. (ie with a .git folder)
```
[the remark this answers is not in the document]
Comment (mikeytdisco) [reply DfciPkg/UnitTests/DfciTests/CloneUnitTests.bat]: Fixed

File: DfciPkg/UnitTests/DfciTests/Support/Python/DFCI_SupportLib.py (hunk 1 above)
Comment (mikeytdisco) [reply DfciPkg/UnitTests/DfciTests/Support/Python/DFCI_SupportLib.py]: Fixed

File: DfciPkg/Application/DfciMenu/DfciCheck429.c (hunk 2 above)
Comment (mikeytdisco) [reply DfciPkg/Application/DfciMenu/DfciCheck429.c:4]: Fixed

File: DfciPkg/Application/DfciMenu/DfciCheck429.c (hunk 3 above)
Comment (mikeytdisco) [reply DfciPkg/Application/DfciMenu/DfciCheck429.c]: Fixed.

File: DfciPkg/Application/DfciMenu/DfciCheck429.c (hunk 4 above)
Comment (mikeytdisco) [reply DfciPkg/Application/DfciMenu/DfciCheck429.c]: Fixed

File: DfciPkg/Application/DfciMenu/DfciCheck429.c (hunk 5 above)
Comment (mikeytdisco) [reply DfciPkg/Application/DfciMenu/DfciCheck429.c]: Fixed

File: DfciPkg/Application/DfciMenu/DfciCheck429.inf (hunk 6 above)
Comment (mikeytdisco) [reply DfciPkg/Application/DfciMenu/DfciCheck429.inf]: Fixed

File: DfciPkg/Application/DfciMenu/DfciCheck429.c [hunk 9]
```diff
@@ -0,0 +1,152 @@
+/** @file
+DfciSARecovery.c
```
Comment (apop5) [inline DfciPkg/Application/DfciMenu/DfciCheck429.c:2]:
```suggestion
DfciSCheck429.c

```

Comment (mikeytdisco) [reply DfciPkg/Application/DfciMenu/DfciCheck429.c]: Fixed for sure this time.

**Pull request merged**